### DISCUSSION TITLE
Extension to manage IWS instances

### DISCUSSION THREAD

**Discussion by worksofliam** [codefori/vscode-ibmi#184] [Ideas]
Microsoft has many extensions to manage Azure services. Wouldn't it be so cool to have an extension for VS Code to manage IWS instances?

There are a bunch of shell scripts to manage services. We just have to make it pretty in vscode!

```sh
bash-4.4$ echo "/QIBM/ProdData/OS/WebServices/bin/listWebServicesServers.sh" | /QOpenSys/usr/bin/qsh
WSERVICE (Stopped)

bash-4.4$ echo "/QIBM/ProdData/OS/WebServices/bin/listWebServices.sh -server WSERVICE" | /QOpenSys/usr/bin/qsh
ConvertTemp (Stopped)

bash-4.4$ echo "/QIBM/ProdData/OS/WebServices/bin/getWebServicesServerProperties.sh -server WSERVICE" | /QOpenSys/usr/bi
n/qsh
Instance path: /www/WSERVICE
Application server: Integrated Application Server 8.5.5.10
Application server ports: 10006;
Subsystem: QHTTPSVR
Job name: WSERVICE
Runtime user ID: QWSERVICE
JVM version: 1.8
JVM type: IBM Technology for Java VM
Web services runtime: Apache CXF 2.6
Web services install path: /www/WSERVICE/webservices/services
Context root: /web/services
Java toolbox tracing: Disabled
Toolbox trace output file name: /www/WSERVICE/wlp/usr/servers/WSERVICE/logs/messages.log
Server log file name: /www/WSERVICE/wlp/usr/servers/WSERVICE/logs/messages.log
HTTP server name: WSERVICE
HTTP server ports: 8333;

bash-4.4$ echo "/QIBM/ProdData/OS/WebServices/bin/getWebServiceProperties.sh -server WSERVICE -service ConvertTemp" | /Q
OpenSys/usr/bin/qsh
Name:                        ConvertTemp
Description:                 ConvertTemp
Startup type:                Automatic
Type:                        SOAP
Status:                      Stopped
Runtime user ID:             *SERVER
Install path:                /www/WSERVICE/webservices/services/ConvertTemp
Program object path:         /QSYS.LIB/QSYSDIR.LIB/QIWSSAMPLE.SRVPGM
PCML file path:              /www/WSERVICE/webservices/services/ConvertTemp/ConvertTemp.pcml
WSDL target namespace URI:   http://converttemp.wsbeans.iseries/
WSDL file path:              /www/WSERVICE/webservices/services/ConvertTemp/META-INF/ConvertTemp.wsdl
WSDL file generation time:   Wed Apr 09 10:44:37 EDT 2014
WSDL file modification time: Wed Feb 05 18:09:37 EST 2020
Operations:                  converttemp (Enabled); converttemp_XML (Enabled);
Library list:                QSYSDIR;
Library list position:       *LAST
Connection pool properties
  Default CCSID:                     *USERID
  Use maintenance threads:           true
  Cleanup interval (seconds):        300
  Maximum number of connections:     *NOMAX
  Maximum inactivity time (seconds): 3600
  Maximum life time (seconds):       86400
  Maximum use count:                 *NOMAX
  Maximum use time (seconds):        *NOMAX
  Fill count:                        1
Transport metadata:
Transport headers:
Secure transport required:           false
Authentication method:               *NONE

bash-4.4$ echo "/QIBM/ProdData/OS/WebServices/bin/createWebServicesServer.sh" | /QOpenSys/usr/bin/qsh
....

bash-4.4$ echo "/QIBM/ProdData/OS/WebServices/bin/installWebService.sh" | /QOpenSys/usr/bin/qsh
.....
```

I would love some opinions on this. Is it worth the time? Or should it just be left alone?

**Comment by richardschoen** [comment]
Couple of thoughts:
I personally would not use anything IWS specific.  If I can help it I'll never use IWS. 

I would rather see something generic that allows me to run ANY bash command line script and see the results.  Then I can use any command line. 

So that could take a couple forms:

Provide a way to execute a bash command line directly and bring back the results. (Like an ssh command line)

Provide a general list of my own user-defined actions for running commands over the ssh command session not tied to source or objects. I think right now they are tied to objects or source members if I recall.  These would be more operational commands. 

You could wrap something like Jesse's Service Commander.  It does a good job of standardizing startup and shutdown stuff.

Food for thought.

**Reply by worksofliam** [reply]
Thanks for your input @richardschoen.

> I would rather see something generic that allows me to run ANY bash command line script and see the results. Then I can use any command line.

This can already be achieved with custom Actions, although I suppose that doesn't work the intended way because you have to open source to be able to run actions. Maybe that could change but is outside the scope of this discussion.

![image](https://user-images.githubusercontent.com/3708366/121816434-8cb0ad80-cc49-11eb-92a0-a679e0bb75b1.png)

This idea would actually be its own extension (but would depend on Code for IBM i for the connection) and would add a new UI to manage the IWS servers and services. Similar to this, but IWS focused of course: https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-azurefunctions

**Reply by richardschoen** [reply]
So managing IWS instances would be something close sourced to paying customers ?  Just so I understand.

**Reply by worksofliam** [reply]
No, it'd be open-source! In fact, I went ahead and made the repo for it so I can get some of the groundwork done.

https://github.com/halcyon-tech/vscode-ibmi-iws

**Reply by richardschoen** [reply]
So what were you trying to say with this:

"This can already be achieved with custom Actions, although I suppose that doesn't work the intended way because you have to open source to be able to run actions. Maybe that could change but is outside the scope of this discussion."

**Reply by worksofliam** [reply]
I was trying to indicate that you can run custom commands, but only when a file is open in the editor. Maybe the base extension should also add a way where you can run commands without needing an open file.

**Reply by richardschoen** [reply]
That would be useful !

**Reply by worksofliam** [reply]
I'll make an issue for it. Thanks!

**Comment by worksofliam** [comment]
Well, I made a start. I likely won't pick it up anymore until I feel like it will be a worthy investment. Could be pretty cool!

Repo is here: https://github.com/halcyon-tech/vscode-ibmi-iws

![image](https://user-images.githubusercontent.com/3708366/121819388-cccc5c00-cc5a-11eb-95a9-f82a10828309.png)

**Comment by ewarwoowar-git** [comment]
I LOVE this idea!

We are getting seriously into using web services in our shop, hosting them using the IWS, and are creating more and more at an ever-increasing pace as business needs demand, so would love the ability to see and manage the status of existing services.

What would also be great would be the ability to copy/promote a service from one server to another.  We have a customised version of a small green screen utility I found on the net to do this, and it uses the IBM shell scripts, but is a little clunky, but something akin to this would be great.

**Comment by WBSpmichal** [comment]
If the extension was able to also add and modify individual services that would be very nice, since the IWS interface is rather clunky. Apparently an interface rewrite is in the works, but I'll believe it when I see it.
